models/ssh_key.go

```diff
@@ -1030,3 +1058,225 @@ func SearchDeployKeys(repoID int64, keyID int64, fingerprint string) ([]*DeployK
 	}
 	return keys, x.Where(cond).Find(&keys)
 }
+
+// PrincipalKey represents principal key information.
+type PrincipalKey struct{}
+
+// AddPrincipalKey adds new principal to database and authorized_principals file.
+func AddPrincipalKey(ownerID int64, name, content string, loginSourceID int64) (*PublicKey, error) {
+	log.Trace(content)
+
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return nil, err
+	}
+
+	// Key name of same user cannot be duplicated.
+	has, err := sess.
+		Where("content = ? AND name = ? AND type = ?", content, name, KeyTypePrincipal).
+		Get(new(PublicKey))
+	if err != nil {
+		return nil, err
+	} else if has {
+		return nil, ErrKeyAlreadyExist{0, "", content}
+	}
```

Comment:
This code is not ensuring that the Key name of the same user cannot be duplicated.

AFAIU a certificate principal needs to match to one user and one user only. The name any particular user gives to that principal is irrelevant:

```suggestion
	// Key name of same user cannot be duplicated.
	has, err := sess.
		Where("content = ? AND type = ?", content, KeyTypePrincipal).
		Get(new(PublicKey))
	if err != nil {
		return nil, err
	} else if has {
		return nil, ErrKeyAlreadyExist{0, "", content}
	}
```

Reply:
You're right, I handled that in the caller which puts the content in name and content, so it's identical.

line 55 of keys.go
```
	if _, err = models.AddPrincipalKey(ctx.User.ID, content, content, 0); err != nil {
```
 
But I can change it here too if you want